---
title: Preparing the Lohoff biorXiv seqFISH dataset for _ExperimentHub_
author: Jonathan Griffiths
date: 10 November 2020
output:
  BiocStyle::html_document:
    toc_float: true
---

```{r style, echo=FALSE, results='hide', message=FALSE}
library(BiocStyle)
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
```

# Overview

Here we will get the data for Pijuan-Sala et al.'s E8.25 ATAC-seq dataset.
As I haven't been involved in this dataset, for information on their methods please see their paper

# Preparing the processed data

We obtain the processed count data through the `r Biocpkg("BiocFileCache")` framework.
This caches the data locally upon the first download, avoiding the need to repeat the download on subsequent analyses.
We download the data from the provided server:

```{r}
library(BiocFileCache)
url <- "https://content.cruk.cam.ac.uk/jmlab/SpatialMouseAtlas2020/"
bfc <- BiocFileCache("raw_data", ask=FALSE)

readme.path <- bfcrpath(bfc, file.path(url, "README.md"))
rme <- read.table(readme.path, sep = "[")[,1]
rme <- rme[substr(rme, 1, 1) == "-"]
files <- gsub("- ", "", regmatches(rme, regexpr("- [A-Za-z0-9\\._]+", rme)))

paths <- lapply(files, function(x) bfcrpath(bfc, file.path(url, x)))
names(paths) <- files
```

## seqFISH counts

We load in the count data from the `.rds` files:

```{r}
library(Matrix)
counts <- readRDS(paths[["counts.Rds"]])
norm <- readRDS(paths[["exprs.Rds"]]) #should be from the area
```

One gene changed name between Ensembl 92 (the atlas) and now.
We change it back, for consistency.

```{r}
rownames(counts)[which(rownames(counts) == "Cavin3")] = "Prkcdbp"
rownames(norm)[which(rownames(norm) == "Cavin3")] = "Prkcdbp"
```

We also can take the cell metadata and tweak some entries for consistency with our other data types.
We separate a data frame of cell positions from it, as well.

```{r}
meta <- readRDS(paths[["metadata.Rds"]])
#extract umap
umap <- meta[,paste0("UMAP", 1:2)]
meta <- meta[,!grepl("UMAP", colnames(meta))]

names(meta)[names(meta) == "uniqueID"] <- "cell"
names(meta)[names(meta) == "celltype_mapped_refined"] <- "celltype"
names(meta)[names(meta) == "Area"] <- "area"
meta$sample <- gsub("embryo", "", meta$embryo)

cell_pos <- meta[,c(
    which(grepl("^[xy]_global_affine", names(meta))),
    which(names(meta) == "z"))]
names(cell_pos) <- c("x", "y")
meta <- meta[,!grepl("(^[xy]_global_affine|^z)", names(meta))]

#Not strings as factors by default
for(i in seq_len(ncol(meta))){
    if(class(meta[,i]) == "factor")
        meta[,i] = as.character(meta[,i])
}
```

Something something size factors

```{r}
sfs <- colSums(counts) - counts["Xist",]
sfs <- sfs/mean(sfs)
my_norm = Matrix(log2(sweep(counts, 2, sfs, "/")+1), sparse = TRUE)
table(my_norm[my_norm > 0] == norm[my_norm > 0])

mismatches = which(my_norm[my_norm > 0] != norm[my_norm > 0])
table(row(counts)[mismatches])

reNormalise = function(sce_filt) {
  assay(sce_filt, "logcounts") <- logcounts(scater::logNormCounts(
    sce_filt, size_factors = sce_filt$total - counts(sce_filt)["Xist",]))
  return(sce_filt)
}
#area
sfs <- meta$Area/mean(meta$Area)
lc.area <- Matrix(log2(t(t(counts)/sfs) + 1), sparse=TRUE)
table(lc.area[lc.area > 0] == norm[lc.area > 0])

#counts not Xist
cs <- colSums(counts[rownames(counts) != "Xist",])
sfs <- (1/cs)/mean(1/cs)
lc.counts <- Matrix(log2(t(t(counts)/sfs) + 1), sparse=TRUE)
table(lc.counts[lc.counts > 0] == norm[lc.counts > 0])


sfs = edgeR::calcNormFactors(counts)
edgeR.counts <- Matrix(log2(t(t(counts)/sfs) + 1), sparse=TRUE)
table(edgeR.counts[edgeR.counts > 0] == norm[edgeR.counts > 0])
```

We need to make our own rowData here.
We will do so from the existing MGD rowData from the atlas.
We also switch the rownames to Ensembl IDs for consistency with other data in the package.

```{r}
library(MouseGastrulationData)
temp <- EmbryoAtlasData(samples=1)
atlas_rd <- rowData(temp)
rd <- atlas_rd[match(rownames(counts), atlas_rd$SYMBOL),]
rownames(rd) <- rownames(counts) <- rd$ENSEMBL
```

mRNA spots and segmentation paths are a special kind of data for seqFISH experiments.
We can incorporate these into the SCE using `BumpyMatrix` files.

```{r}
library(BumpyMatrix)
spots <- readRDS(paths[["mRNA.Rds"]])
spots$geneID <- as.character(spots$geneID)
spots$geneID[spots$geneID == "Cavin3"] <- "Prkcdbp"
names(spots)[3:4] <- c("x", "y")
spots$ENSEMBL <- rd$ENSEMBL[match(spots$geneID, rd$SYMBOL)]

spots_bm <- splitAsBumpyMatrix(
    x=spots[, !names(spots)%in%c("ENSEMBL", "uniqueID", "geneID")],
    row=spots$ENSEMBL,
    col=spots$uniqueID,
    sparse=TRUE #there are many 0-counts
)
#order correctly and remove QC fail cells
spots_bm <- spots_bm[rownames(counts), colnames(counts)]
```

We also rearrange the segmentation coordinates into a DataFrameList

```{r}
library(IRanges)
dfl = DataFrameList(lapply(seq_len(nrow(meta)), function(x) 
    data.frame(
        x = meta$segmentation_vertices_x_global_affine[x][[1]],
        y = meta$segmentation_vertices_y_global_affine[x][[1]]
    )))
meta <- meta[,!grepl("segmentation_vertices_[xy]_global_affine", names(meta))]
meta <- DataFrame(meta)
meta$segmentation_vertices = dfl
```

Now we make the observed count SCE.

```{r}
library(SpatialExperiment)
sce <- SpatialExperiment(
    assays = List(
        counts = counts,
        spots = spots_bm
    ),
    colData = meta,
    rowData = rd,
    reducedDims = List(umap = umap),
    spatialCoords = cell_pos
)
sizeFactors(sce) <- sfs
```

We now save the data, splitting the large `SingleCellExperiment` object into its constituent parts.
We then upload these smaller files to `r Biocpkg("ExperimentHub")`.
Splitting up the data allows us to update the various getter functions if _SingleCellExperiment_ is updated or overhauled.

```{r}
base <- file.path("MouseGastrulationData", "lohoff_seqfish_biorxiv", "1.6.0")
dir.create(base, recursive=TRUE, showWarnings=FALSE)
saveRDS(rowData(sce), file=paste0(base, "/rowdata.rds"))
for(samp in unique(sce$sample)){
    sub = sce[, sce$sample == samp]
    saveRDS(counts(sub), 
        file=paste0(base, "/counts-processed-sample", samp, ".rds"))
    saveRDS(assays(sub, "spots"), 
        file=paste0(base, "/spots-processed-sample", samp, ".rds"))
    saveRDS(colData(sub), 
        file=paste0(base, "/coldata-sample", samp, ".rds"))
    saveRDS(sizeFactors(sub), 
        file=paste0(base, "/sizefac-sample", samp, ".rds"))
    saveRDS(reducedDims(sub), 
        file=paste0(base, "/reduced-dims-sample", samp, ".rds"))
    saveRDS(spatialCoords(sub), 
        file=paste0(base, "/spatial-coords-sample", samp, ".rds"))
}
```

## Imputed counts

The imputed counts are transcriptome wide, but are fairly dense compared to true scRNA-seq data.
We load and organise this data for saving.

```{r}
library(rhdf5)
imp <- h5read(paths[["imputed.h5"]], "logcounts")
imp <- Matrix(imp, sparse=TRUE) #reduces memory size somewhat
colnames(imp) <- readRDS(paths[["imputed_column_names.Rds"]])
rn_imp <- readRDS(paths[["imputed_row_names.Rds"]])
```

Note that there are some genes where the symbol is duplicated, with one example shown below.

```{r}
rn_imp[grepl("Nudt8", rn_imp)]
```

We can show that, if we remove the suffixes, we perfectly recapitulate the atlas gene symbols.

```{r}
rd_imp <- data.frame(
    ENSEMBL=atlas_rd$ENSEMBL[match(rn_imp, atlas_rd$SYMBOL)],
    SYMBOL=rn_imp)
ens <- atlas_rd$ENSEMBL[match(rn_imp, atlas_rd$SYMBOL)]
mismatch <- which(is.na(ens))
table(gsub("\\.[12]$", "", rn_imp[mismatch]) == atlas_rd$SYMBOL[mismatch])
```

Moreover, if we ignore the `r length(mismatch)` mismatching genes, we can see that every other one matches the atlas rowData.

```{r}
match <- which(!is.na(ens))
table(rn_imp[match] == atlas_rd$SYMBOL[match])
```

In addition, pers. comm.s confirmed that the genes are ordered the same.
Altogether, we take the atlas rowData forward for this dataset.

```{r}
rownames(imp) = atlas_rd$ENSEMBL
sce_imp <- SpatialExperiment(
    assays = List(
        logcounts_imputed = imp[,meta$cell]
    ),
    colData = meta,
    rowData = atlas_rd,
    spatialCoords = cell_pos
)
```

Now we can save this, split by sample

```{r}
base <- file.path("MouseGastrulationData", "lohoff_seqfish_biorxiv_imputed", "1.6.0")
dir.create(base, recursive=TRUE, showWarnings=FALSE)
saveRDS(rowData(sce_imp), file=paste0(base, "/rowdata.rds"))
for(samp in unique(sce_imp$sample)){
    sub = sce_imp[, sce_imp$sample == samp]
    saveRDS(logcounts(sub), 
        file=paste0(base, "/logcounts-processed-sample", samp, ".rds"))
    saveRDS(colData(sub), 
        file=paste0(base, "/coldata-sample", samp, ".rds"))
    saveRDS(spatialCoords(sub), 
        file=paste0(base, "/spatial-coords-sample", samp, ".rds"))
}
```

# Make file metadata

We now make the metadata for ExperimentHub so the files can be made properly available.

```{r}
info <- data.frame(
    Title = sprintf("BPS_atac %s",
        c(sprintf("processed counts (sample %i)", 1),
            "rowData",
            sprintf("colData (sample %i)", 1),
            sprintf("size factors (sample %i)", 1),
            sprintf("reduced dimensions (sample %i)", 1),
            sprintf("raw counts (sample %i)", 1))
    ),
    Description = sprintf("%s for the E8.25 mouse embryo single-cell ATAC-seq dataset", 
        c(sprintf("Processed counts for sample %i", 1),
            "Per-gene metadata for all samples",
            sprintf("Per-cell metadata for sample %i", 1),
            sprintf("Size factors for sample %i", 1),
            sprintf("Reduced dimensions for sample %i", 1),
            sprintf("Raw (unfiltered) counts for sample %i", 1))
    ),
    RDataPath = c(
        file.path("MouseGastrulationData", "BPS_atac", "1.6.0", 
            c(sprintf("counts-processed-sample%i.rds", 1),
                "rowdata.rds",
                sprintf("coldata-sample%i.rds", 1),
                sprintf("sizefac-sample%i.rds", 1),
                sprintf("reduced-dims-sample%i.rds", 1),
                sprintf("counts-raw-sample%i.rds", 1)))
    ),
    BiocVersion="3.13",
    Genome="mm10",
    SourceType="TXT",
    SourceUrl=c(
        "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE133244",
        rep("https://static-content.springer.com/", 2),
        "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE133244",
        "https://static-content.springer.com/",
        "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE133244"
    ),
    SourceVersion=c(
        "GSE133244_embryo_revision1_allPeaks_afterClusterPeak_raw.mtx.gz",
        rep("41556_2020_489_MOESM2_ESM.xlsx", 2),
        "GSE133244_embryo_revision1_allPeaks_afterClusterPeak_raw.mtx.gz",
        "41556_2020_489_MOESM2_ESM.xlsx",
        "GSE133244_embryo_revision1_allPeaks_afterClusterPeak_raw.mtx.gz"
    ),
    Species="Mus musculus",
    TaxonomyId="10090",
    Coordinate_1_based=FALSE,
    DataProvider="Jonathan Griffiths",
    Maintainer="Jonathan Griffiths <jonathan.griffiths.94@gmail.com>",
    RDataClass="character",
    DispatchClass="Rds",
    stringsAsFactors = FALSE
)

write.csv(file="../../extdata/metadata-bpsatac.csv", info, row.names=FALSE)

```

# Session information

```{r}
sessionInfo()
```
