---
title: Preparing the _Tal1_ chimera dataset for _ExperimentHub_
author: Aaron Lun
date: 28 May 2019
output:
  BiocStyle::html_document:
    toc_float: true
---

```{r style, echo=FALSE, results='hide', message=FALSE}
library(BiocStyle)
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
```

# Overview

Here, we prepare a single-cell RNA sequencing dataset that studies the effect of _Tal1_ knock-out in early mouse embryogenesis.
This dataset contains four 10X Genomics samples, with two replicates each for the wild-type and _Tal1_ chimeric conditions.
We will set up both the raw count matrices from _CellRanger_ as well as a highly processed form of the data.
For the latter, full details of the processing can be found at https://github.com/MarioniLab/EmbryoTimecourse2018.

# Preparing the processed data

We obtain the processed count data through the `r Biocpkg("BiocFileCache")` framework.
This caches the data locally upon the first download, avoiding the need to repeat the download on subsequent analyses.

```{r}
library(BiocFileCache)
bfc <- BiocFileCache("raw_data", ask=FALSE)
count.path <- bfcrpath(bfc, file.path("https://content.cruk.cam.ac.uk/",
    "jmlab/chimera_tal1_data/raw_counts.mtx.gz"))
```

We load in the count data from the MatrixMarket format, using methods from the `r CRANpkg("Matrix")` package:

```{r}
library(Matrix)
counts <- readMM(count.path)
dim(counts)
```

We download the cell- and gene-level metadata using `r Biocpkg("BiocFileCache")`, and read them into R.

```{r}
meta.path <- bfcrpath(bfc, file.path("https://content.cruk.cam.ac.uk/",
    "jmlab/chimera_tal1_data/meta.tab.gz"))
meta.tab <- read.delim(meta.path, stringsAsFactors=FALSE)
head(meta.tab)

gene.path <- bfcrpath(bfc, file.path("https://content.cruk.cam.ac.uk/",
    "jmlab/chimera_tal1_data/genes.tsv.gz"))
gene.tab <- read.delim(gene.path, header=FALSE, stringsAsFactors=FALSE)
colnames(gene.tab) <- c("ENSEMBL", "SYMBOL")
head(gene.tab)
```

We store the count matrix in a `SingleCellExperiment` object with the metadata associated with each cell and gene.

```{r}
library(SingleCellExperiment)
sce <- SingleCellExperiment(list(counts=counts), 
    colData=meta.tab, rowData=gene.tab)
sce
```

We also obtain the size factors and store them in `sce`.

```{r}
sf <- bfcrpath(bfc, file.path("https://content.cruk.cam.ac.uk/",
    "jmlab/chimera_tal1_data/sizefactors.tab.gz"))
sf <- read.delim(sf, header=FALSE, stringsAsFactors=FALSE)[,1]
sizeFactors(sce) <- sf
head(sf)
```

We save each component of the `SingleCellExperiment` to file for upload to `r Biocpkg("ExperimentHub")`.

```{r}
base <- file.path("MouseEarlyEmbryoData", "tal1-chimera", "1.0.0")
dir.create(base, recursive=TRUE, showWarnings=FALSE)
saveRDS(counts(sce), file=file.path(base, "counts-processed-all.rds"))
saveRDS(rowData(sce), file=file.path(base, "rowdata.rds"))
saveRDS(colData(sce), file=file.path(base, "coldata.rds"))
saveRDS(sizeFactors(sce), file=file.path(base, "sizefac.rds"))
```

# Setting up the raw data

We obtain the raw count data through the `r Biocpkg("BiocFileCache")` framework.
Each file contains the raw (unfiltered) count matrix from _CellRanger_ for each sample.

```{r}
sample.paths <- character(4)
for (i in seq_along(sample.paths)) {
    fname <- sprintf("sample_%s_unswapped.mtx.gz", i)
    sample.paths[i] <- bfcrpath(bfc, 
        file.path("https://content.cruk.cam.ac.uk/",
       "jmlab/chimera_tal1_data/unfiltered", fname))
}
```

We read in each sparse matrix and serialize it into a sample-specific RDS file.

```{r}
collected <- vector("list", length(sample.paths))
for (i in seq_along(collected)) {
    curmat <- readMM(sample.paths[i])
    saveRDS(curmat, file=file.path(base, sprintf("counts-raw-sample%i,rds", i)))
}
```

Note that the row-level metadata is the same in both the raw and processed data, and does not need to be re-acquired.

# Session information

```{r}
sessionInfo()
```

